# Convert the read counts of multiple samples generated by featureCounts program
# from Subread software, according to specified experiment design file, based on
# the assumption:
#
# The names of 3'-DGE sequence alignment files have a form of:
#
# RNAseq_[Date].[Well].umi.[sam|bam]
#
# Therefore, the columns of read counts matrix can be match with the rows of
# experiment design table by the following steps:
#
# 1) Extract the Well number from the names of sequence alignment files (included
#    in the header row of the read counts files generated by featureCounts), and
#    assign the Well number to corresponding sample columns of read counts matrix.
#
# 2) Ensure all the Well numbers from the experiment design table are included in
#    those from the read counts matrix.
#
# 3) Extract and sort the columns of the read counts matrix according to the rows
#    of the experiment design table by the Well numbers of samples.

# Load required library
library("tools")

# User home directory
user_home <- Sys.getenv("HOME")

# Control parameters.
# Data text type.
data_text_type <- "txt"
# Data table type.
data_table_type <- "tsv"
# Flag of verbose print.
verbose <- TRUE

# Directory for dataset.
dataset_dir <- file.path(user_home, "LINCSData/Datasets/Alignment/LINCS.Dataset/iPSC-CM/DGE-GEO-Depot")
stopifnot(dir.exists(dataset_dir))
# Directory for read counts tables of dataset.
counts_dir <- file.path(dataset_dir, "Counts")
stopifnot(dir.exists(counts_dir))
# Name pattern of sequence alignment files.
align_file_name_ptrn <- "\\.(sam|bam)$"

# Read the experiment design of multiple samples.
exprt_design_file_dir <- counts_dir
stopifnot(dir.exists(exprt_design_file_dir))
exprt_design_file_main_name <- "DGE-RNAseq-Experiment-Design"
exprt_design_file_name <- paste(exprt_design_file_main_name, data_table_type, sep=".")
exprt_design_file_path <- file.path(exprt_design_file_dir, exprt_design_file_name)
stopifnot(file.exists(exprt_design_file_path))
if(verbose) cat("Reading ", exprt_design_file_path, "...\n", sep="")
exprt_design <- read.delim(exprt_design_file_path, quote="", comment.char="#", check.names=FALSE, stringsAsFactors=FALSE)
exprt_design$ID <- paste(exprt_design$State, exprt_design$iPSC.Line, exprt_design$Experiment, exprt_design$Dish, sep=".")
stopifnot(!any(duplicated(exprt_design$ID)))

# Process sample names and sort sample columns.
read_counts_file_main_name <- "DGE-RNAseq-Read-Counts"
read_counts_file_name <- paste(read_counts_file_main_name, data_text_type, sep=".")
read_counts_file_dir <- counts_dir
read_counts_file_path <- file.path(read_counts_file_dir, read_counts_file_name)
stopifnot(file.exists(read_counts_file_path))

# Read read counts table.
if(verbose) cat("Reading ", read_counts_file_path, "...\n", sep="")
read_counts <- read.delim(read_counts_file_path, quote="", comment.char="#", check.names=FALSE, stringsAsFactors=FALSE)

# Move Geneid column to row names.
rownames(read_counts) <- read_counts$Geneid
read_counts$Geneid <- NULL

# Extract read counts columns whose names contain "bam" keyword.
read_counts <- read_counts[,grepl(align_file_name_ptrn,colnames(read_counts),ignore.case=TRUE)]

# Clean up sample names by removing leading directory and extract [Well].
sample_names <- basename(colnames(read_counts))
colnames(read_counts) <- sapply(strsplit(sample_names,split="\\."), function(x){return(paste0(x[2],collapse="."))})

# Extract and sort samples according to experiment design and gene names.
stopifnot(all(exprt_design$Well %in% colnames(read_counts)))
read_counts_subset <- read_counts[,exprt_design$Well]
read_counts <- read_counts[order(rownames(read_counts)),]

# Write converted read counts of selected samples.
convert_read_counts_file_dir <- counts_dir
convert_read_counts_file_name <- paste(read_counts_file_main_name, data_table_type, sep=".")
convert_read_counts_file_path <- file.path(convert_read_counts_file_dir, convert_read_counts_file_name)
if(verbose) cat("Writing ", convert_read_counts_file_path, "...\n", sep="")
write.table(read_counts, file=convert_read_counts_file_path, quote=FALSE, na="", sep="\t", row.names=TRUE, col.names=TRUE)
