# Comparing gene/protein expression profiles to identify differentially
# expressed molecules (DEM), such as differentially expressed genes (DEG)
# or differentially expressed proteins (DEP).

# Convert a field name to one of the field names of merged experiment design table.
convertExperimentDesignFieldName <- function(field_name, orig_exprt_design_field_names, merged_exprt_design_field_names, field_names_positions, std_exprt_design_field_names, exprt_design_field_name_map)
{
    field_name_cvt <-NULL
    if(field_name %in% orig_exprt_design_field_names)
    {
        if(!(field_name %in% merged_exprt_design_field_names))
        {
            field_name_idx <- match(field_name, orig_exprt_design_field_names)
            field_name_position_flags <- field_names_positions == field_name_idx
            if(any(field_name_position_flags))
            {
                field_name_std <- names(field_names_positions)[field_name_position_flags]
                if(field_name_std %in% names(exprt_design_field_name_map)) field_name_cvt <- exprt_design_field_name_map[field_name_std]
                else field_name_cvt <- field_name_std
            }
            else warning("field name must be one of the field names of origial experiment design table!")
        }
        else field_name_cvt <-field_name
    }
    else
    {
        if(field_name %in% std_exprt_design_field_names)
        {
            std_exprt_design_field_names_mapped <- std_exprt_design_field_names
            names(std_exprt_design_field_names_mapped) <- std_exprt_design_field_names
            std_exprt_design_field_names_mapped[names(exprt_design_field_name_map)] <- exprt_design_field_name_map
            field_name_cvt <- std_exprt_design_field_names_mapped[field_name]
        }
        else warning("field name must be one of the eight standardized field names of experiment design table!")
    }
    # Double check the availability of batch field name in merged experiment
    # design table to be used in calculation.
    if(!is.null(field_name_cvt) && !(field_name_cvt %in% merged_exprt_design_field_names)) warning("Failed to convert field name to any of the field names of merged experiment design table!")
    # Return the converted field name with NULL for failed conversion.
    return(field_name_cvt)
}

# Comparing gene/protein expression profiles to identify differentially
# expressed molecules (DEM), such as differentially expressed genes (DEG)
# or differentially expressed proteins (DEP).
#
# Input datasets for comparison are the raw read counts of mRNA or peptide
# expression levels, measured at different drug-treated conditions as well
# as control condition and generated by RNA-sequencing or protein-spectrum
# experiments,

compareMoleculeExpression <- function(func_dir, repo_dir, config_file, read_counts_title="ReadCounts", exp_configs_title="ExpDesigns", deg_title="DEG", top_title="TOP", calc_title="Calc", plot_title="Plots")
{
    # Turn on warning output.
    orig_warn_status <- getOption("warn")
    options(warn=1)

    ####################### Load required function library #######################

    # Load required library
    require(matrixStats)
    require(tools)
    # These R function scripts need to be in the same directory as this main script.
    source(file.path(func_dir, "openPlotDevice.R"), local=TRUE)
    source(file.path(func_dir, "closePlotDevice.R"), local=TRUE)
    source(file.path(func_dir, "initializeParameters.R"), local=TRUE)
    source(file.path(func_dir, "replaceIllegalChars.R"), local=TRUE)
    source(file.path(func_dir, "readReadCountsDatasets.R"), local=TRUE)
    source(file.path(func_dir, "mergeReadCountsDatasets.R"), local=TRUE)
    source(file.path(func_dir, "adaptReadCountsDatasets.R"), local=TRUE)
    source(file.path(func_dir, "sumExperimentDesign.R"), local=TRUE)
    source(file.path(func_dir, "filterGeneExpSamples.R"), local=TRUE)
    source(file.path(func_dir, "plotHeatMapOfGeneExpCorDist.R"), local=TRUE)
    source(file.path(func_dir, "calcExpressionDifference.R"), local=TRUE)

    ############ Initialize various parameters for comparison analysis ############

    # Set data diretories.
    counts_dir <- file.path(repo_dir, "Counts")
    params_dir <- file.path(repo_dir, "Params")
    scripts_dir <- file.path(repo_dir, "Scripts")
    results_dir <- file.path(repo_dir, "Results")
    repo_dirs <- c(counts_dir, params_dir, scripts_dir, results_dir)
    repo_dirs_avail <- dir.exists(repo_dirs)
    if(!all(repo_dirs_avail)) stop(paste(paste0(repo_dirs[!repo_dirs_avail],collapse=" and "), "doesn't exist!"))

    # Initializing various input parameters for differential expression analysis.
    initializeParameters(config_file=config_file, counts_dir=counts_dir, params_dir=params_dir, scripts_dir=scripts_dir, func_dir=func_dir)

    # Define a set of eight standard field names for experiment design table.
    # - Species:   biological species of a sample column.
    # - Subject:   species subject of a sample column.
    # - State:     biological or perturbation condition of a sample column.
    # - Culture:   culture code of a sample column.
    # - Replicate: replicate code of a sample column.
    # - Measure:   measure code of a sample column.
    # - Sample:    sample code of a sample column.
    # - Time:      time point of a sample column
    std_exprt_design_field_names <- c("Species", "Subject", "State", "Culture", "Replicate", "Measure", "Sample", "Time")

    # Check field_names_positions and field_types.
    stopifnot(setequal(names(field_names_positions), names(field_types)))
    stopifnot(setequal(names(field_names_positions), std_exprt_design_field_names))

    ############# Read in read counts data experiment design schemes #############

    # Step 1. Read in read counts datasets and experiment design schemes.
    # Note: an ID field will be created in experiment design table and will be
    # assigned to the columns of read counts table.
    read_counts_datasets_original <- readReadCountsDatasets(read_counts_files=read_counts_files, generateReadCountsColumnNames=generateReadCountsColumnNames, split_char=split_char, field_names=field_names, field_types=field_types, std_exprt_design_field_names=std_exprt_design_field_names, species_default=species_default, subject_default=subject_default, state_default=state_default, culture_default=culture_default, measure_default=measure_default, time_default=time_default, row_name_column=row_name_column, read_counts_columns=read_counts_columns, read_counts_skip=read_counts_skip, exprt_design_files=exprt_design_files, field_names_positions=field_names_positions, exprt_design_skip=exprt_design_skip, legal_chars=legal_chars, repl_chars=repl_chars, compress_space=compress_space, func_dir=func_dir)

    # Step 2. Perform gene/protein-specific processing of read counts datasets.
    if(!is.null(processReadCountsDatasets))
    {
        processReadCountsDatasets <- match.fun(processReadCountsDatasets)
        stopifnot(is.function(processReadCountsDatasets))
        read_counts_datasets_original <- processReadCountsDatasets(read_counts_datasets_original)
    }

    # Step 3. Merge multiple read counts matrices and experiment design tables.
    # Note: read_counts_datasets_merged may contain all-zero rows.
    read_counts_datasets_merged <- mergeReadCountsDatasets(read_counts_datasets=read_counts_datasets_original, merge_method=read_counts_merge_method, missing_value=missing_value, func_dir=func_dir)

    # Assign merged read counts matrix and experiment design table to separate
    # variables for compatibility with subsequent comparison analysis.
    read_counts_merged <- read_counts_datasets_merged$read_counts
    exprt_design_merged <- read_counts_datasets_merged$exprt_design
    orig_exprt_design_field_names <- read_counts_datasets_merged$orig_exprt_design_field_names

    # Step 4. Adapt some eof the standard field names of design design table to the
    # variable names used in the source codes of comparison analysis which were
    # developed before the field names became standardized. The adaptation rules
    # include:
    #  Subject --> Cell
    #  Culture --> Experiment
    #  Replicate --> Dish
    #  Measure --> Plate
    #  Sample --> Well
    exprt_design_field_name_map <- c(Subject="Cell", Culture="Experiment", Replicate="Dish", Measure="Plate", Sample="Well")
    stopifnot(all(names(exprt_design_field_name_map) %in% std_exprt_design_field_names))
    exprt_design_merged <- adaptReadCountsDatasets(exprt_design=exprt_design_merged, field_name_map=exprt_design_field_name_map)

    # Match subset field name to the field name of experiment design table.
    if(!is.null(subset_field_name))
    {
        subset_field_name <- convertExperimentDesignFieldName(field_name=subset_field_name, orig_exprt_design_field_names=orig_exprt_design_field_names, merged_exprt_design_field_names=colnames(exprt_design_merged), field_names_positions=field_names_positions, std_exprt_design_field_names=std_exprt_design_field_names, exprt_design_field_name_map=exprt_design_field_name_map)
        if(is.null(subset_field_name)) stop("Failed to convert subset field name to any of the field names of merged experiment design table!")
    }

    # Match batch field name to the field name of experiment design table.
    if(!is.null(batch_field_name))
    {
        batch_field_name <- convertExperimentDesignFieldName(field_name=batch_field_name, orig_exprt_design_field_names=orig_exprt_design_field_names, merged_exprt_design_field_names=colnames(exprt_design_merged), field_names_positions=field_names_positions, std_exprt_design_field_names=std_exprt_design_field_names, exprt_design_field_name_map=exprt_design_field_name_map)
        if(is.null(batch_field_name)) stop("Failed to convert batch field name to any of the field names of merged experiment design table!")
    }
    # Ensure subset and batch field names are not the same.
    if(!(is.null(subset_field_name) || is.null(batch_field_name))) stopifnot(subset_field_name != batch_field_name)

    # Select eight standard fields plus ID, subset and batch fields for experiment
    # design table used in differential comparison computation.
    comp_exprt_design_field_names <- std_exprt_design_field_names
    names(comp_exprt_design_field_names) <- std_exprt_design_field_names
    comp_exprt_design_field_names[names(exprt_design_field_name_map)] <- exprt_design_field_name_map
    comp_exprt_design_field_names <- c(comp_exprt_design_field_names, "ID")
    names(comp_exprt_design_field_names) <- NULL
    # Add subset and batch field names as needed.
    if((!is.null(subset_field_name)) && (!(subset_field_name%in%comp_exprt_design_field_names))) comp_exprt_design_field_names <- c(comp_exprt_design_field_names, subset_field_name)
    if((!is.null(batch_field_name)) && (!(batch_field_name%in%comp_exprt_design_field_names))) comp_exprt_design_field_names <- c(comp_exprt_design_field_names, batch_field_name)
    exprt_design_merged <- exprt_design_merged[,comp_exprt_design_field_names]
    # Save modified experiment design table.
    read_counts_datasets_merged$exprt_design <- exprt_design_merged

    # Misc data customization.
    # Change all CTRL-related state name to CTRL.
    exprt_design_merged$State[grepl(state_name_ctrl, exprt_design_merged$State, ignore.case=TRUE)] <- state_name_ctrl
    # Sort exprt_design_merged by Cell, State, and Experiment.
    exprt_design_merged <- exprt_design_merged[order(xtfrm(exprt_design_merged[,"Species"]),xtfrm(exprt_design_merged[,"Cell"]),xtfrm(exprt_design_merged[,"State"]),xtfrm(exprt_design_merged[,"Experiment"]),xtfrm(exprt_design_merged[,"Dish"]),xtfrm(exprt_design_merged[,"Plate"]),xtfrm(exprt_design_merged[,"Well"])),]
    # Sort sample columns in read counts table according to the order of
    # the samples in experiment design.
    read_counts_merged <- read_counts_merged[,exprt_design_merged$ID]

    ############# Post process read counts data experiment design schemes #############

    # Set plate number to 0 for merging sample replicates across different plates.
    if(merge_plates)
    {
        # Set the data type of merged plate number.
        measure_field_type <- field_types["Measure"]
        if(measure_field_type == "character") merged_plate_name <- as.character(merged_plate_name)
        else if(measure_field_type == "numeric") merged_plate_name <- as.numeric(merged_plate_name)
        else if(measure_field_type == "logical") merged_plate_name <- as.logical(merged_plate_name)
        else merged_plate_name <- as.character(merged_plate_name)
        # Save original plate number.
        exprt_design_merged[[orig_plate_field_name]] <- exprt_design_merged$Plate
        # Set merged plate number.
        exprt_design_merged$Plate <- merged_plate_name
    }

    # Set the name of cell species and time points.
    # Only allow single cell species and time point in differential expression
    # comparison, because the compariosns involving multiple cell species and
    # time points needs to be set up in different ways.
    bio_spec <- unique(exprt_design_merged$Species)
    stopifnot(length(bio_spec)==1)
    time_point <- unique(exprt_design_merged$Time)
    stopifnot(length(time_point)==1)
    time_length <- paste("Hour", time_point, sep="-")

    # Save the filtered exprt design table and read counts.
    exprt_design_merged_file <- file.path(results_dir, "Experiment-Design-Merged.tsv")
    write.table(exprt_design_merged, file=exprt_design_merged_file, quote=FALSE, sep="\t", row.names=FALSE)
    # Write the read counts to a data file.
    read_counts_merged_file <- file.path(results_dir, "Read-Counts-Merged.tsv")
    write.table(read_counts_merged, file=read_counts_merged_file, quote=FALSE, sep="\t")

    #################### Summarize original experiment design ####################

    state_cell_plate_rep <- sumExperimentDesign(exprt_design_merged)
    if(!is.null(state_cell_plate_rep))
    {
        drug_sample_reps <- data.frame(No=rownames(state_cell_plate_rep), state_cell_plate_rep, stringsAsFactors=FALSE)
        colnames(drug_sample_reps) <- c("No", "Drug", "Cell", "Plate", "Rep")
        drug_sample_reps_file <- file.path(results_dir, "Drug-Sample-Reps-Original.tsv")
        write.table(drug_sample_reps, file=drug_sample_reps_file, sep="\t", quote=FALSE, row.names=FALSE)
        if(verbose1) print("Original Drug Sample Replicates:")
        if(verbose1) print(state_cell_plate_rep)
    }

    ####################### Read in pre-defined drug groups #######################

    # The file of drug groups represents a global drug-group config for
    # all merged experiments.
    drug_groups_table <- read.delim(file=drug_groups_file, quote="", comment.char="#", stringsAsFactors=FALSE)
    drug_groups <- list()
    drug_groups[[state_name_ctrl]] <- state_name_ctrl
    # Replcae pre-defined illegal characters in drug names with spaces.
    if(!is.null(legal_chars) && is.character(legal_chars) && nchar(legal_chars)>0) drug_groups_table$Drug <- replaceIllegalChars(string=drug_groups_table$Drug, legal_chars=legal_chars, repl_chars=repl_chars, compress_space=compress_space)
    # Assert no duplicated drug names in drug_groups.
    stopifnot(all(!duplicated(drug_groups_table$Drug)))
    # A drug group contains multiple drugs belonging to this group and the group
    # name is retrieved from drug group table.
    for(drug_group in drug_groups_table$Group) drug_groups[[drug_group]] <- drug_groups_table[drug_groups_table$Group==drug_group,"Drug"]

    # Compare the pre-defined drug groups with available treatment conditions in
    # exprt design table (experiment design) to detect if there are any mismatches.
    # And it's possible that not all pre-defined drug names of each drug group can
    # be matched with available drug-treated conditions in each cell line.
    drug_names <- unique(unlist(drug_groups, use.names=FALSE))
    state_names <- unique(exprt_design_merged$State)
    if(verbose1) print(paste("Pre-defined drug groups and available experiment conditions are", (if(setequal(drug_names, state_names)) "the same" else "different")))
    diff_drug_state <- setdiff(drug_names, state_names)
    if(length(diff_drug_state) > 0)
    {
        if(verbose1) print("The drugs defined in drug groups but unavaible in experiment conditions:")
        if(verbose1) print(diff_drug_state)
    }
    diff_state_drug <- setdiff(state_names, drug_names)
    if(length(diff_state_drug) > 0)
    {
        if(verbose1) print("The drugs avaible in experiment conditions but undefined in drug groups:")
        if(verbose1) print(diff_state_drug)
    }

    ###################### Remove outlier sample replicates ######################

    # Filter out outlier samples according to divisive clustering analysis, such that
    # any samples which depart from the rest samples by greater than dist_cutoff_outlier
    # (for group with more than min_samples number of samples) or dist_cutoff_group are
    # removed (for group with min_samples number of samples).

    # Load customized cutoff values for outlier samples.
    outlier_cutoffs <- read.delim(file=outlier_cutoffs_file, quote="", comment.char="#", stringsAsFactors=FALSE)
    # Replcae pre-defined illegal characters in state names with spaces.
    if(!is.null(legal_chars) && is.character(legal_chars) && nchar(legal_chars)>0) outlier_cutoffs$State <- replaceIllegalChars(string=outlier_cutoffs$State, legal_chars=legal_chars, repl_chars=repl_chars, compress_space=compress_space)

    # Prepare PDF file for plotting.
    if(plot_clust)
    {
        clust_image_plot_flow_by_row <- TRUE
        if(!is.null(clust_image_plot_flow))
        {
            clust_image_plot_flow <- tolower(clust_image_plot_flow)
            if(clust_image_plot_flow=="row" || clust_image_plot_flow=="column")
            {
                if(clust_image_plot_flow == "row") clust_image_plot_flow_by_row <- TRUE
                else clust_image_plot_flow_by_row <- FALSE
            }
            else stop("clust_image_plot_flow must be either row or column!")
        }
        else stop("clust_image_plot_flow cannot be NULL!")
        clust_main_title <- paste(bio_spec, time_length, sep="-")
        clust_plot_file_name <- paste(clust_main_title, "Cluster-Plots", sep="-")
        dev_info <- openPlotDevice(dev_type=clust_image_type, file_name=clust_plot_file_name, dir_name=results_dir, width=clust_image_width, height=clust_image_height, single_file=single_clust_image_file)
        orig_par <- par(no.readonly = TRUE)
        layout(matrix(c(1:(clust_image_rows*clust_image_columns)),nrow=clust_image_rows,ncol=clust_image_columns,byrow=clust_image_plot_flow_by_row), respect=clust_image_keep_aspect_ratio)
        par(mar=c(bottom=6, left=5.5, top=5.5, right=2) + 0.1)
    }

    # Determine color_field_pos according to color_field_name.
    sample_id_fields <- read_counts_datasets_merged$sample_id_fields
    sample_id_field_names <- names(sample_id_fields)
    if(!(color_field_name %in% sample_id_field_names)) stop(paste0("color_field_name must be one of ", paste0(sample_id_field_names,collapse=", "), "!"))
    color_field_pos <- sample_id_fields[color_field_name]
    if(!(color_field_type=="character" || color_field_type=="numeric")) stop("color_field_type must be either character or numeric!")

    # Set drug names.
    drug_names <- unique(unlist(drug_groups, use.names=FALSE))

    # Sort and filter sample replicates in each drug-treated condition for
    # all drug groups and cell lines.
    exprt_design_merged <- filterGeneExpSamples(exprt_design_merged=exprt_design_merged, read_counts_merged=read_counts_merged, drug_names=drug_names, dist_cutoffs=outlier_cutoffs, subset_field_name=subset_field_name, dist_cutoff_outlier=dist_cutoff_outlier, dist_cutoff_group=dist_cutoff_group, min_samples=min_samples, filter_outlier=filter_outlier, plot_orig_clust=plot_clust, plot_filter_clust=plot_clust, plot_pass_clust=plot_clust, plot_empty_clust=(plot_clust&&plot_empty_clust), color_sample_groups=color_sample_groups, color_field_pos=color_field_pos, color_field_type=color_field_type, clust_title_elems_names=clust_title_elems_names, clust_title_elems_flags=clust_title_elems_flags, clust_title_elems_keys=clust_title_elems_keys, leg_title=color_text_name, verbose=(if(verbose1) verbose2 else verbose1), func_dir=func_dir)
    read_counts_merged <- read_counts_merged[,exprt_design_merged$ID]
    # Remove the rows containing all zero read counts due to the removal of outlier samples.
    read_counts_merged <- read_counts_merged[rowSums(read_counts_merged)>0,]
    # Save filtered experiment design and read counts datasets.
    read_counts_datasets_filtered <- list(exprt_design=exprt_design_merged, read_counts=read_counts_merged)

    # Save the filtered exprt design table and read counts.
    exprt_design_filtered_file <- file.path(results_dir, "Experiment-Design-Filtered.tsv")
    write.table(exprt_design_merged, file=exprt_design_filtered_file, quote=FALSE, sep="\t", row.names=FALSE)
    # Write the read counts to a data file.
    read_counts_filtered_file <- file.path(results_dir, "Read-Counts-Filtered.tsv")
    write.table(read_counts_merged, file=read_counts_filtered_file, quote=FALSE, sep="\t")

    # Restore orginal graphics settings and close the plot.
    if(plot_clust)
    {
        par(orig_par)
        closePlotDevice(dev_info)
    }

    #################### Summarize filtered experiment design ####################

    state_cell_plate_rep_filter <- sumExperimentDesign(exprt_design_merged)
    if(!is.null(state_cell_plate_rep_filter))
    {
        drug_sample_reps <- data.frame(No=rownames(state_cell_plate_rep_filter), state_cell_plate_rep_filter, stringsAsFactors=FALSE)
        colnames(drug_sample_reps) <- c("No", "Drug", "Cell", "Plate", "Rep")
        drug_sample_reps_filter_file <- file.path(results_dir, "Drug-Sample-Reps-Filtered.tsv")
        write.table(drug_sample_reps, file=drug_sample_reps_filter_file, sep="\t", quote=FALSE, row.names=FALSE)
        if(verbose1) print("Filtered Drug Sample Replicates:")
        if(verbose1) print(state_cell_plate_rep_filter)
    }

    # Check the availability of exprt_design_merged after outlier removal.
    if(is.null(exprt_design_merged) || (!is.null(exprt_design_merged) && nrow(exprt_design_merged)==0)) stop("Experiment design after outlier removal is empty!")

    #################### Plot the heat map of sample distance ####################

    if(plot_heatmap)
    {
        # Prepare PDF file for plotting.
        heatmap_main_title <- paste(bio_spec, time_length, sep="-")
        heatmap_plot_file_name <- paste(heatmap_main_title, "Distance-Plots", sep="-")
        dev_info <- openPlotDevice(dev_type=heatmap_image_type, file_name=heatmap_plot_file_name, dir_name=results_dir, width=heatmap_image_width, height=heatmap_image_height, single_file=single_heatmap_image_file)
        orig_par <- par(no.readonly = TRUE)

        # Plot the drugs of each groups across all cell lines.
        for(drug_group_name in names(drug_groups))
        {
            drug_group <- drug_groups[[drug_group_name]]
            for(state_name in drug_group)
            {
                if(verbose1) print(paste("Drug", state_name, "in", drug_group_name, "group across cell lines."))
                # Prepare the datasets.
                exprt_design_merged_state <- exprt_design_merged[exprt_design_merged$State==state_name,]
                # Some drugs have been removed by outlier-removal step.
                if(nrow(exprt_design_merged_state) < 2) next
                wellplate_names_state <- exprt_design_merged_state$ID
                state_names_state <- exprt_design_merged_state$State
                cell_names_state <- exprt_design_merged_state$Cell
                read_counts_state <- read_counts_merged[,wellplate_names_state]
                read_counts_state <- read_counts_state[rowSums(is.na(read_counts_state))==0,]
                if(is.vector(read_counts_state)) read_counts_state <- as.matrix(read_counts_state)
                # Plot heat map of distance matrix.
                cell_names_str <- paste0(unique(cell_names_state),collapse="/")
                title_text <- paste(ncol(read_counts_state), state_name, "Samples of", drug_group_name, "Group in Cell", cell_names_str)
                plotHeatMapOfGeneExpCorDist(gene_exp=read_counts_state, group_names=state_names_state, use_group_names=FALSE, title_text=title_text, dist_method="correlation", norm_factors=FALSE, lower_limit=lower_limit, upper_limit=upper_limit, margin_bottom=11.5, margin_right=11.5, func_dir=func_dir)
            }
        }

        # Close the PDF plot.
        par(orig_par)
        closePlotDevice(dev_info)
    }

    ####### Compare read counts data and obtain differentially expressed molecules ##########

    # Determine whether or not to generate various data files.
    if(!write_read_counts_file) read_counts_title <- NULL
    if(!write_exp_configs_file) exp_configs_title <- NULL
    if(!write_deg_file) deg_title <- NULL
    if(!write_top_file) top_title <- NULL
    if(!write_calc_file) calc_title <- NULL
    if(!write_plot_file) plot_title <- NULL

    # Compare the difference of molecular expression levels for genes and proteins, and
    # return both raw read counts and differential comparison statistics of all molecules.
    # Note: only the drug treatments included in both pre-defined drug groups and experiment
    # design table will be used for differential comparison.
    deg_counts_stats <- calcExpressionDifference(read_counts=read_counts_merged, exprt_design=exprt_design_merged, drug_groups=drug_groups, bio_spec=bio_spec, time_length=time_length, results_dir=results_dir, subset_field_name=subset_field_name, batch_field_name=batch_field_name, ctrl_state_name=state_name_ctrl, use_design_matrix=use_design_matrix, use_glm_ql=use_glm_ql, fdrs=fdrs, min_samples=min_samples, min_samples_req=min_samples_req, dispersion=dispersion, n_top=n_top, deg_only=deg_only, norm_base=norm_base, remove_small=remove_small, read_counts_title=read_counts_title, deg_title=deg_title, top_title=top_title, exp_configs_title=exp_configs_title, calc_title=calc_title, plot_title=plot_title, deg_image_type=deg_image_type, deg_image_width=deg_image_width, deg_image_height=deg_image_height, single_deg_image_file=single_deg_image_file, plot_bcv=plot_bcv, plot_smear=plot_smear, pt_chars=pt_chars, verbose1=verbose1, verbose2=verbose2, func_dir=func_dir)

    ################################## Clean up ##################################

    # Restore warning output.
    options(warn=orig_warn_status)

    # Save and return multiple datasets of read counts and experiment design.
    # Original: multiple original molecular read counts
    # Merged: the merged molecular read counts from multiple original read counts.
    # Filtered: the filtered molecular read counts after the removal of outlier samples and all-zero read counts.
    read_counts_datasets <- list(original=read_counts_datasets_original, merged=read_counts_datasets_merged, filtered=read_counts_datasets_filtered, degstats=deg_counts_stats)
    # Save the dataset object to a RData file.
    if(!is.null(read_counts_datasets_rdata_file))
    {
        if(tolower(file_ext(read_counts_datasets_rdata_file)) != "rdata") read_counts_datasets_rdata_file <- paste(read_counts_datasets_rdata_file, "RData", sep=".")
        read_counts_datasets_rdata_file <- file.path(results_dir, read_counts_datasets_rdata_file)
        save(read_counts_datasets, file=read_counts_datasets_rdata_file)
    }

    return(read_counts_datasets)
}
